### PR TITLE
Metaprogramming facilities to support hacky_wrapper to reorder function arguments

### PR DESCRIPTION
**Pull request by smessmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49164 Remove codegen logic to support non-c10-full ops
* #49138 [wip] Making ops c10-full: list of optional tensors
* #49150 Making ops c10-full: ConstQuantizerPtr
* #49146 Making ops c10-full: Storage arguments
* #49088 Making ops c10-full: optional lists
* #49083 Making ops c10 full: optional out arguments
* #49013 Making ops c10-full: Generator arguments
* #49012 Making ops c10 full: out overloads with default arguments
* #49008 Making ops c10-full: Dimname arguments
* #49007 Making ops c10-full: ops blocked by manual registrations
* #49006 Making ops c10 full: ops returning multiple out arguments
* #48912 Make out ops c10-full (with hacky-wrapper)
* #48911 hacky_wrapper_for_legacy_signatures reorders out arguments
* **#48910 Metaprogramming facilities to support hacky_wrapper to reorder function arguments**
* #49079 remove hacky_wrapper from BackendSelect
* #47793 Make aten::div.out c10-full

This doesn't implement any reordering of arguments in hacky_wrapper yet, but it implements the metaprogramming facilities needed for that.
A PR on top of this one will implement the actual hacky_wrapper reordering.

Differential Revision: [D25363337](https://our.internmc.facebook.com/intern/diff/D25363337/)

### CODE REVIEW

File: c10/util/IntegerSequence.h [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+using drop_t = typelist_to_iseq_t<typelist::drop_t<iseq_to_typelist_t<ISeq>, N>>;
+
+/**
+ * Make an index_sequence representing a range of values.
+ * Begin is inclusive, end is exclusive.
+ * Example:
+ *  range_t<3, 8> == std::index_sequence<3, 4, 5, 6, 7>
+ */
+template<size_t begin, size_t end>
+struct range {
```
Comment (bhosmer) [inline c10/util/IntegerSequence.h:97]: This duplicates some of the functionality in [Metaprogramming.h](https://github.com/pytorch/pytorch/blob/master/c10/util/Metaprogramming.h#L134) - we should probably only have one implementation of this concept. (Also note that the implementation there doesn't require generating the list prefix, which I could imagine mattering in certain use cases.)

**Review by bhosmer — COMMENTED** [review]
Hey, not done but thought I'd flush the comment queue before I context switched away from this - back with more tmrw 😁

**Comment by ezyang** [comment]
My high level impression on this PR is that, if we had implemented hacky wrapper as codegen, we wouldn't need any of these metaprogramming utilities, we wouldn't need to backdoor mutable argument locations as an integer sequence into the template in question, and it would be simpler to understand too.

Now, I understand hacky wrapper is supposed to be temporary, so I don't want to litigate this too much. So most of my concerns boil down to (1) short term necessity of reviewing these PRs (@bhosmer, maybe you are OK with fielding the reviews here?) and (2) long term maintenance of this metaprogramming facilities (@smessmer has expressed openness to deleting them when hacky wrapper dies, or investigating adding a proper metaprogramming library like Hana as a dep to PyTorch.)

**Comment by bhosmer** [comment]
> My high level impression on this PR is that, if we had implemented hacky wrapper as codegen, we wouldn't need any of these metaprogramming utilities, we wouldn't need to backdoor mutable argument locations as an integer sequence into the template in question, and it would be simpler to understand too.
> 
> Now, I understand hacky wrapper is supposed to be temporary, so I don't want to litigate this too much. So most of my concerns boil down to (1) short term necessity of reviewing these PRs (@bhosmer, maybe you are OK with fielding the reviews here?) and (2) long term maintenance of this metaprogramming facilities (@smessmer has expressed openness to deleting them when hacky wrapper dies, or investigating adding a proper metaprogramming library like Hana as a dep to PyTorch.)

Yup I'll be reviewing this part of the stack. Template complexity is definitely a concern, @smessmer I'll sync up offline to chat about degrees of freedom here

**Review by bhosmer — COMMENTED** [review]
Given that the lifecycle for this is the same as hacky_wrapper, and we want to go to something like Hana for longer-lived metaprogramming support, I'd like to try to prune away anything here that's not strictly needed to implement #48911. 

What that amounts to depends on whether arbitrary out argument patterns need to be supported - if so, I think we can significantly simplify this, and if not, we might be able to reduce it to a couple of helpers.

In either case, IMO the lifting of ints to types is overkill - the functionality we're reusing from typelist isn't significant enough to warrant the indirection; we should just implement the simple list primitives we need over integer sequences directly.

So next step is resolving the supported-patterns question in #48911, then back to here for a revisit.

**Review by bhosmer — APPROVED** [review]

**Comment by bhosmer** [comment]
Hey, following up on the accept: after looking at #48911, I'd suggest seriously considering making the change I proposed there, which would make this PR unnecessary.

**Comment by smessmer** [comment]
This PR got so small now that I could merge it into https://github.com/pytorch/pytorch/pull/48911

**Pull request closed without merging**